File: server/lib/game_machine/app_config.rb
require 'singleton'
require_relative 'hocon_config'
module GameMachine
  class AppConfig
    include Singleton

    attr_reader :config, :loaded

    def initialize
      @loaded = false
      @config = OpenStruct.new
    end

    def load_config
      @config = HoconConfig.config
      set_config_from_env

      set_java_config
      @loaded = true
    end

    def set_java_config
      JavaLib::AppConfig.set_env_root(ENV['APP_ROOT'])
      JavaLib::AppConfig.set_env(GameMachine.env)
      JavaLib::AppConfig.set_default_game_id(config.default_game_id)
      JavaLib::AppConfig.set_orm(config.orm)
      JavaLib::AppConfig.set_plugins(config.plugins)
      JavaLib::AppConfig::Handlers.setAuth(config.handlers.auth)
      JavaLib::AppConfig::Gamecloud.set_host(config.gamecloud.host)
      JavaLib::AppConfig::Gamecloud.set_user(config.gamecloud.user)
      JavaLib::AppConfig::Gamecloud.set_api_key(config.gamecloud.api_key)

      JavaLib::AppConfig::Jdbc.set_hostname(config.jdbc.hostname)
      JavaLib::AppConfig::Jdbc.set_port(config.jdbc.port)
      JavaLib::AppConfig::Jdbc.set_database(config.jdbc.database)
      JavaLib::AppConfig::Jdbc.set_username(config.jdbc.username)
      JavaLib::AppConfig::Jdbc.set_password(config.jdbc.password)
      JavaLib::AppConfig::Jdbc.set_ds(config.jdbc.ds)
      JavaLib::AppConfig::Jdbc.set_driver(config.jdbc.driver)
      JavaLib::AppConfig::Jdbc.set_url(config.jdbc.url)
      
      JavaLib::AppConfig::Couchbase.set_bucket(config.couchbase.bucket)
      JavaLib::AppConfig::Couchbase.set_password(config.couchbase.password)
      JavaLib::AppConfig::Couchbase.set_servers(config.couchbase.servers)
      JavaLib::AppConfig::Datastore.set_store(config.datastore.store)
      JavaLib::AppConfig::Datastore.set_serialization(config.datastore.serialization)
      JavaLib::AppConfig::Datastore.set_cache_write_interval(config.datastore.cache_write_interval)
      JavaLib::AppConfig::Datastore.set_cache_writes_per_second(config.datastore.cache_writes_per_second)
      JavaLib::AppConfig::Datastore.set_mapdb_path(config.datastore.mapdb_path)

      JavaLib::AppConfig::set_game_config(config.game)
            
      JavaLib::AppConfig::Client.setIdleTimeout(config.client.idle_timeout)
    end

    def set_config_from_env
      GameMachine.logger.info "Setting config from ENV"

      if ENV['NODE_HOST']
        config.http.host = ENV['NODE_HOST']
        config.tcp.host = ENV['NODE_HOST']
        config.udp.host = ENV['NODE_HOST']
        config.akka.host = ENV['NODE_HOST']
      end

      if ENV['WWW_PORT']
        config.http.port = ENV['WWW_PORT'].to_i
      end

      if ENV['TCP_PORT']
        config.tcp.port = ENV['TCP_PORT'].to_i
      end

      if ENV['UDP_PORT']
        config.udp.port = ENV['UDP_PORT'].to_i
      end

      if ENV['AKKA_PORT']
        config.akka.port = ENV['AKKA_PORT'].to_i
      end

      if ENV['CLOUD_HOST'] && ENV['CLOUD_USER'] && ENV['API_KEY']
        GameMachine.logger.info "Found gamecloud config in ENV"
        config.gamecloud.host = ENV['CLOUD_HOST']
        config.gamecloud.user = ENV['CLOUD_USER']
        config.gamecloud.api_key = ENV['API_KEY']
      end

      if ENV['DB_HOST'] && ENV['DB_PORT']  && ENV['DB_NAME'] && ENV['DB_USER'] && ENV['DB_PASS']
        GameMachine.logger.info "Found database config in ENV"
        config.jdbc.hostname =  ENV['DB_HOST']
        config.jdbc.port =      ENV['DB_PORT'].to_i
        config.jdbc.database =  ENV['DB_NAME']
        config.jdbc.username =  ENV['DB_USER']
        config.jdbc.password =  ENV['DB_PASS']
      end

      if ENV['AKKA_SEED_HOST'] && ENV['AKKA_SEED_PORT']
        GameMachine.logger.info "Found Seed in ENV"
        Application.config.seeds << "#{ENV['AKKA_SEED_HOST']}:#{ENV['AKKA_SEED_PORT']}"
      end

    end

    private

  end
end

File: server/java/project/src/main/java/com/game_machine/core/AppConfig.java
package com.game_machine.core;

import java.util.List;


public class AppConfig {
	
	private static String env;

    public static class Datastore {
    	private static String store;
    	private static String serialization;
    	private static long cacheWriteInterval;
    	private static long cacheWritesPerSecond;
    	
		public static String getStore() {
			return store;
		}
		public static void setStore(String store) {
			Datastore.store = store;
		}
		public static String getSerialization() {
			return serialization;
		}
		public static void setSerialization(String serialization) {
			Datastore.serialization = serialization;
		}
		public static long getCacheWriteInterval() {
			return cacheWriteInterval;
		}
		public static void setCacheWriteInterval(long cacheWriteInterval) {
			Datastore.cacheWriteInterval = cacheWriteInterval;
		}
		public static long getCacheWritesPerSecond() {
			return cacheWritesPerSecond;
		}
		public static void setCacheWritesPerSecond(long cacheWritesPerSecond) {
			Datastore.cacheWritesPerSecond = cacheWritesPerSecond;
		}
		
    }
    
	public static class Couchbase {
		private static String bucket;
		private static String password;
		private static List<String> servers;
		public static String getBucket() {
			return bucket;
		}
		public static void setBucket(String bucket) {
			Couchbase.bucket = bucket;
		}
		public static String getPassword() {
			return password;
		}
		public static void setPassword(String password) {
			Couchbase.password = password;
		}
		public static List<String> getServers() {
			return servers;
		}
		public static void setServers(List<String> servers) {
			Couchbase.servers = servers;
		}
	}
	
	
	public static class Jdbc {
		private static String hostname;
		private static int port;
		private static String database;
		private static String ds;
		private static String driver;
		private static String username;
		private static String password;
		
		
		public static String getHostname() {
			return hostname;
		}
		public static void setHostname(String hostname) {
			Jdbc.hostname = hostname;
		}
		public static int getPort() {
			return port;
		}
		public static void setPort(int port) {
			Jdbc.port = port;
		}
		public static String getDatabase() {
			return database;
		}
		public static void setDatabase(String database) {
			Jdbc.database = database;
		}
		public static String getDs() {
			return ds;
		}
		public static void setDs(String ds) {
			Jdbc.ds = ds;
		}
		public static String getDriver() {
			return driver;
		}
		public static void setDriver(String driver) {
			Jdbc.driver = driver;
		}
		public static String getUsername() {
			return username;
		}
		public static void setUsername(String username) {
			Jdbc.username = username;
		}
		public static String getPassword() {
			return password;
		}
		public static void setPassword(String password) {
			Jdbc.password = password;
		}
	}
	
	public static class Gamecloud {
		private static String host;
		private static String user;
		private static String apiKey;
		
		public static String getHost() {
			return host;
		}
		public static void setHost(String host) {
			Gamecloud.host = host;
		}
		public static String getUser() {
			return user;
		}
		public static void setUser(String user) {
			Gamecloud.user = user;
		}
		public static String getApiKey() {
			return apiKey;
		}
		public static void setApiKey(String apiKey) {
			Gamecloud.apiKey = apiKey;
		}
	}
	
	public static class Handlers {
		
		private static String auth;
		private static String team;
		
		
		public static String getAuth() {
			return auth;
		}
		public static void setAuth(String auth) {
			Handlers.auth = auth;
		}
		public static String getTeam() {
			return team;
		}
		public static void setTeam(String team) {
			Handlers.team = team;
		}
	}

	public static String getEnv() {
		return env;
	}

	public static void setEnv(String env) {
		AppConfig.env = env;
	}
}
